src/codegate/pipeline/pii/pii.py

```diff
@@ -0,0 +1,345 @@
+
+import re
+from typing import Any, Dict, List, Optional
+
+import structlog
+from litellm import ChatCompletionRequest, ChatCompletionSystemMessage, ModelResponse
+from litellm.types.utils import Delta, StreamingChoices
+
+from codegate.config import Config
+from codegate.pipeline.base import (
+    PipelineContext,
+    PipelineResult,
+    PipelineStep,
+)
+from codegate.pipeline.output import OutputPipelineContext, OutputPipelineStep
+from codegate.pipeline.pii.manager import PiiManager
+from codegate.pipeline.systemmsg import add_or_update_system_message
+
+logger = structlog.get_logger("codegate")
+
+
+class CodegatePii(PipelineStep):
+    """
+    CodegatePii is a pipeline step that handles the detection and redaction of PII
+    in chat completion requests.
+
+    Methods:
+        __init__:
+            Initializes the CodegatePii pipeline step and sets up the PiiManager.
+
+        name:
+            Returns the name of the pipeline step.
+
+        _get_redacted_snippet(message: str, pii_details: List[Dict[str, Any]]) -> str:
+            Extracts a snippet of the message containing all detected PII.
+
+        process(request: ChatCompletionRequest, context: PipelineContext) -> PipelineResult:
+            Processes the chat completion request to detect and redact PII. Updates the request with
+            anonymized text and stores PII details in the context metadata.
+
+        restore_pii(anonymized_text: str) -> str:
+            Restores the original PII from the anonymized text using the PiiManager.
+    """
+
+    def __init__(self):
+        """Initialize the CodegatePii pipeline step."""
+        super().__init__()
+        self.pii_manager = PiiManager()
```

Comment:
Do you think we need to initialize a New `PiiManager` every time we initialize the step?

Maybe it is confusing and we should refactor the code, but the step's `__init__` is eventually called for every request. For secrets, we instantiate a `SecretsManager` outside the pipeline and then pass it in when we create a pipeline instance.

Reply:
btw I know there there is an issue about extracting the common code for the secret and PII obfuscation - we can solve my comment with that. I was just trying to understand if it was intentional to initialize a new instance of PiiManager per request.

Reply:
Yup, this needs fixing, as pipeline factory expects SecretsManager:

```
class PipelineFactory:
    def __init__(self, secrets_manager: SecretsManager):
    ```
 
Just working on changing this.